File: Python/beginner/1827.py
# https://judge.beecrowd.com/pt/problems/view/1827
# Matriz Quadrdada IV

# Loop to read n until EOF
while True:
    try:
        n = int(input())
    except EOFError:
        break
    except ValueError:
        continue
    
    startInner = n // 3
    endInner = n - 1 - (n // 3)
    mid = n // 2
    
    # Loop to print the matrix
    for rows in range(n):
        for columns in range(n):
            # Central Point. Needs to be found first
            if rows == mid and columns == mid:
                print(4, end='')
            # Inner Matrix
            elif rows >= startInner and rows <= endInner and columns >= startInner and columns <= endInner: 
                print(1, end='')
            # Main Diagonal
            elif rows == columns: 
                print(2, end='')
            # Secondary Diagonal    
            elif rows + columns == n - 1: 
                print(3, end='')
            # Other positions
            else: 
                print(0, end='')
        print()
    print()
# By: Fernando Serra
# https://github.com/fernandosserra

File: Haskell/beginner/1827.hs
-- https://judge.beecrowd.com/pt/problems/view/1827
-- Matriz Quadrada IV

-- Importing replicateM and printf
import Control.Monad (replicateM)

-- Function to calculate Central Point, Inner Matrix, Main and Sec. Diagonals and External Values
positionFinder :: Int -> Int -> Int -> Int
positionFinder r c n =
    let startInner = n `div` 3
        endInner = n - 1 - (n `div` 3)
        mid = n `div` 2
    in
    if r == mid && c == mid then 4 -- Central Point
    else if r >= startInner && r <= endInner && c >= startInner && c <= endInner then 1 -- Inner Matrix
    else if r == c then 2  -- Main diagonal 
    else if r + c == n - 1 then 3  -- Secondary diagonal (needs to be evaluated first)
    else 0  -- Other positions

-- Main Function
main :: IO ()
main = do
    -- Loop to read multiple entries until EOF
    contents <- getContents
    mapM_ processLine $ filter (not . null) $ lines contents
  where
    processLine line = do
        let n = read line :: Int
        -- Generating our Matrix line by line
        let matrix = [[positionFinder r c n | c <- [0..n-1]] | r <- [0..n-1]]
        -- Printing the matrix
        mapM_ (putStrLn . concatMap show) matrix
        putStrLn "" -- a new line after each matrix

-- By: Fernando Serra and the all-knowing Washu Hakubi, the ultimate Haskell master!
-- https://github.com/fernandosserra